### ISSUE TITLE
Caching

### ISSUE THREAD

**Issue by Erikvv** [ikod/dlang-requests#81]
I'm implementing a cache on top of this library which uses the Cache-Control, Expires and ETag headers. Like a web browser does.

I wonder if the core contributers (1) would like to have this included (2) have any considerations on the design.

Specifically considering the design
- Should the cache be global or something to be passed around by the library user?
- The cache needs to be accessible by multiple threads. Should this be done with locks or with message passing?

**Comment by Erikvv** [comment]
It would be nice to have interceptors to implement this feature. Like in Axios or angular/common/http

**Comment by ikod** [comment]
Hello Eric!

Can you please provide any link regarding interception? I'd like to read
any docs on it as I never have deal with angular.

Thanks!

пт, 31 авг. 2018 г., 3:09 Erik van Velzen <notifications@github.com>:

**Comment by Erikvv** [comment]
It's built on top of ReactiveX. It may be hard to understand if you have never used ReactiveX. 

Very popular is using 2 interceptors for authorization:

https://medium.com/@ryanchenkie_40935/angular-authentication-using-the-http-client-and-http-interceptors-2f9d1540eb8

Simple interceptor for caching:

https://nrempel.com/guides/angular-httpclient-httpinterceptor-cache-requests/

If you look at HTTP servers there exist similar interfaces. Like PHP's PSR-15 MiddlewareInterface. This receives a callback for the next handler which it can decide to call and do extra processing or circumvent entirely. 

There are alternative like Guzzle's event system but personally I like the more functional approaches.

**Comment by Erikvv** [comment]
It is difficult to add interceptors to this library because there is no separation between a HTTPClient and a HTTPRequest like in most libraries. 

Is that a conscious choice?

**Comment by ikod** [comment]
Hello, @Erikvv 

Do you want to have some hook in the very beginning of the request handling, so that you can return response object from cache, right?

Or another scenario - you add some interceptor and it add/remove/modify some request headers before any other processing.

This is idea? I don't think it is too difficult.
For example we can add code like this (pseudocode)

```
Response get(url) {
     if ( interceptors ) {
        (modified_request, ready response) = interceptors.handle(this)
        if ( ready_response) {
          return ready_response;
         }
        // otherwise continue with modified request
     }
     ... - usual processing continue
```

**Comment by Erikvv** [comment]
The interceptor also needs to be able to access the response, to store it and get the eTag from 304 response.

I will write a proposal.

**Comment by Erikvv** [comment]
No worries doesn't look like I have time right now anyways

**Comment by ikod** [comment]
hello @Erikvv 
Before I finish this change, I'd like to clear out one thing.

```
    Response handle(Request request) {
        if (this.queue.empty) {
            throw new Exception("none of the interceptors finished the request");
        }

        auto interceptor = this.queue.front;
        this.queue.popFront();

        return interceptor(request, this);
    }
```
As Request is `struct` this code means that each interceptor receive, modify and pass down it's own copy of `Request`. This is fine, because initial request stay immutable. On the other hand - each `Request` has some private caches (permanent redirect locations, connection cache) which are filled during request execution. In case each interceptor receive its own copy of Request and pass it down all these caches are stay invisible and actually unusable. 

I see two options here - 
* pass ref to `Request` - not very good as I'd like do not change initial request.
* copy these caches to `Response` so that they can 'move back' to initial Request.

Sorry it this sounds weird. What do you think?

Thanks for your reply.

**Comment by Erikvv** [comment]
High-level almost all http libraries have 3 things:

1. Client
2. Request
3. Response

Angular has HttpClient, HttpResponse, HttpRequest
Java has HttpClient, HttpResponse, HttpRequest
Guzzle has GuzzleHttp\Client, GuzzleHttp\Psr7\Request, GuzzleHttp\Psr7\Response
Python Requests has Session, Response, and Request

I think Request and Response should be entirely immutable. 
All the persistent state you mention should be in Client/Session. Like you mention redirects, persistent connections. But also the interceptors themselves!

In basis, if you don't use Client/Session, you can't add your own interceptors. Though if you look at Python they added a second parameter for this like `requests.get('https://httpbin.org/', hooks={'response': print_url})`.

By the way you could also implement redirect as an interceptor.

On top of this there is sugar so you get say `getContent("http://www.example.com")` without worrying about Request and Client. In that case you make the Client locally and destroy it immediately afterwards (or you use some globally cached Client).

**Comment by Erikvv** [comment]
I understand this would be a big change and I understand if you reject it.

What I don't understand is why there is no Client in the first place. I asked it before --> https://github.com/ikod/dlang-requests/issues/81#issuecomment-417784170

**Comment by ikod** [comment]
Actually `Request` is `Client`, so I just have to move some persistent data to it from HTTPRequest and pass HTTPRequest to interceptors instead of Request. What do you think?

**Comment by Erikvv** [comment]
I'm taking a look now

**Comment by ikod** [comment]
Hello, @Erikvv 

Thanks! I pushed next part of changes. There is lot of code refactoring in comparison with master, but all tests works ok. There is minor changes in API (I deprecated direct access to HTTPRequest and also changed a bit how streamed requests works).

I'm also would like add module-wide interception, so, that library user can install/uninstall interceptors without changing too much code. These interceptors will work with every Request() without explicit calls to Request.addInterceptor()

**Comment by ikod** [comment]
Hello @Erikvv 

Added "global" interceptors, here is how it works:
https://github.com/ikod/dlang-requests/blob/issue_81_2/source/requests/request.d#L610
```
    class PathModifier : Interceptor {
        Response opCall(Request r, RequestHandler next)
        {
            r.path = r.path ~ "get";
            auto rs = next.handle(r);
            return rs;
        }
    }
    // add PathMoifier to  static_interceptors
    addInterceptor(new PathModifier());
    // from now any request will pass through Pathmodifier without changing your code:
    rq = Request();
    rs = rq.get("http://httpbin.org/");
    assert(rs.uri.path == "/get", "Expected /get, but got %s".format(rs.uri.path));

    rs = rq.post("http://httpbin.org/post", "abc"); // real request will be to http://httpbin.org/postget
    assert(rs.code == 404);
}
```

**Comment by ikod** [comment]
Hello, @Erikvv 

I merged changes to master

**Comment by ikod** [comment]
Hello @Erikvv 

I released v1.0.0 with interceptors, so I'll close this issue.

Feel free to reopen it any time
Thanks!
